### PR TITLE
Uwsgi gevent support

### PR DESCRIPTION
**Pull request by efficiosoft**
I added websocket support for gevent with uWSGI using uWSGI's native websocket api. There is currently no detailed documentation for this update, but code comments should suffice for you to decide whether to merge this into master or not. Feedback is heavily appreciated. Best regards.

### CODE REVIEW

**Comment by efficiosoft** [comment]
Sorry for these travis-ci build failures. I'm currently adjusting the tests accordingly.

**Comment by efficiosoft** [comment]
Ok, tests fixed now and all travis-ci checks pass. Sorry again for that delay. Would love to hear your feedback.

**Comment by miguelgrinberg** [comment]
This is awesome, thanks a lot. I will test in the next few days and let you know if I find any problems.

File: engineio/async_gevent.py [hunk 1]
```diff
@@ -52,12 +56,71 @@ def wait(self):
...
+            return None
+        return msg.decode()
+
+    def send_and_wait(self, poll_func):
+        """Calls poll_func(block=False) regularly and sends the messages
+        it returns. Apart from that, it observes the websocket for incoming
+        messages. If one arrives, it returns that. After the returned message
+        was processed, this function can be called again to continue its work.
+        A return of None means that the connection was closed. Afterwards."""
+        while True:
```
Comment (miguelgrinberg) [inline engineio/async_gevent.py]: I'm going to test this, but from the looks of it, this is going to consume CPU, whether there is data traffic or not.  I prefer the approach I'm using, which lets the I/O functions block in another thread until there is something to do.

**Comment by efficiosoft** [comment]
@miguelgrinberg Yes, this is preferable. But what I didn't mention: uWSGI websocket api can only be used from the main greenlet. In another project I used gevent.select on both the connection's and a redis's file descriptor that was used to sync events between servers. But since we have a JoinableQueue here instead of redis, I'm not sure whether select can be used. Maybe one can get the file descriptor of the JoinableQueue's pipe and call select on that?

**Comment by efficiosoft** [comment]
BTW: JoinableQueue._reader works with select.select(), but it's a private attribute and thus not part of the official multiprocessing api. However, I could change the implementation to work with it if you prefer. But I'm not sure about how this would perform on Windows because .fileno() doesn't return a file descriptor there.

**Comment by efficiosoft** [comment]
I committed a new approach that lacks the socket polling at the cost of an extra greenlet and an extra queue per websocket. Should be reasonable I think.

**Comment by miguelgrinberg** [comment]
Thanks, this looks much better. :)

Do you have a link to where uWSGI documents the limitations with calling the websocket functions from a greenlet that is not the main one? I could not find any mention of that in the docs.

**Comment by efficiosoft** [comment]
I couldn't find it in the docs either, but their github issues are full of "[BUG] current_wsgi_req NOT FOUND !!!".
This is valid for the whole uwsgi python api, including functions like uwsgi.connection_fd(). Unfortunately, the documentation for the uwsgi python module is less than rare, but I could reproduce the current_wsgi_req not found bug mentioned above.

Some resources mentioning it:
https://github.com/unbit/uwsgi/issues/980
https://github.com/unbit/uwsgi/blob/master/plugins/gevent/gevent.c

But however their documentation is not the best, uwsgi is one of the most performant and stable wsgi container servers.

**Comment by efficiosoft** [comment]
This version now dynamically chooses the most performant approach depending on the uWSGI version available. I'll probably leave it this way. Tested it with stable (2.0.13.1) and development (2.1-dev-907dbd5). Both work.

**Comment by efficiosoft** [comment]
@miguelgrinberg I'm still unsure about the encoding to pass to msg.decode() on messages received over websocket. uWSGI apparently provides them as bytes, but engineio expects them to be of type str. AFAIK websockets use UTF-8 when operated in non-binary mode So I guess I should specify 'utf-8' explicitly when decoding received messages, right?
On the other hand, engineio passes str objects into send(). uWSGI's send function accepts them, but probably they should be encoded with UTF-8 as well before passing them to uWSGI for transmission?

**Comment by efficiosoft** [comment]
Thoughts about encoding and decoding are cleared now. Binary support is included and working. Please let me know if there is something else I could implement or have to test.

**Comment by miguelgrinberg** [comment]
@efficiosoft what's the uwsgi command line that you use to start the server?

**Comment by efficiosoft** [comment]
This is my current uwsgi.yaml:

```
uwsgi:
    gevent: 1000
    processes: 1
    listen: 128
    http-socket: 127.0.0.1:8080
    static-map: /static=./static
    static-index: index.html
    module: myapp.wsgi
    logformat: %(status) - %(method) %(uri) - %(msecs) ms
    harakiri-verbose: 1
    no-orphans: 1
    vacuum: 1
    master: 1
    enable-threads: 1
    touch-reload: ./uwsgi.reload
```

**Comment by miguelgrinberg** [comment]
Okay, got it to work with the following command line:

    uwsgi --http :5000 --wsgi-file latency.py --callable app --gevent 1000 --http-websockets --master

Thanks.

**Comment by miguelgrinberg** [comment]
Seems to work well, so I just merged these changes, after I made a bit of restructuring. The main difference is that I moved the gevent+uwsgi support to a separate async mode, so now have to specify `async_mode='gevent_uwsgi'`. With this change, in most cases you don't even need to explicitly set the async mode, as it will be selected automatically if the correct dependencies are installed.

**Comment by efficiosoft** [comment]
Thanks a lot for merging! Splitting it into two modules seems reasonable. I updated the doc string.for `Server` class because it didn't reflect the new `async_mode` yet. You might want to merge this as well.

**Pull request closed without merging**